x/gamm/genesis_test.go

```diff
@@ -0,0 +1,133 @@
+	require.NoError(t, err)
+	require.Equal(t, pool.GetId(), poolStored.GetId())
+	require.Equal(t, pool.GetAddress(), poolStored.GetAddress())
+	require.Equal(t, pool.GetPoolParams(), poolStored.GetPoolParams())
+	require.Equal(t, pool.GetTotalWeight(), poolStored.GetTotalWeight())
+	require.Equal(t, pool.GetTotalShares(), poolStored.GetTotalShares())
+	require.Equal(t, pool.GetAllPoolAssets(), poolStored.GetAllPoolAssets())
+	require.Equal(t, pool.String(), poolStored.String())
+
+	_, err = app.GAMMKeeper.GetPool(ctx, 2)
+	require.Error(t, err)
+
+	liquidity := app.GAMMKeeper.GetTotalLiquidity(ctx)
+	require.Equal(t, liquidity, sdk.Coins{sdk.NewInt64Coin("nodetoken", 10), sdk.NewInt64Coin(sdk.DefaultBondDenom, 10)})
+}
+
+func TestGammExportGenesis(t *testing.T) {
+	app := simapp.Setup(false)
+	ctx := app.BaseApp.NewContext(false, tmproto.Header{})
+
+	acc1 := sdk.AccAddress(ed25519.GenPrivKey().PubKey().Address().Bytes())
+	app.BankKeeper.SetBalances(ctx, acc1, sdk.Coins{
+		sdk.NewInt64Coin("foo", 100000),
+		sdk.NewInt64Coin("bar", 100000),
+	})
+
+	_, err := app.GAMMKeeper.CreatePool(ctx, acc1, types.PoolParams{
+		SwapFee: sdk.NewDecWithPrec(1, 2),
+		ExitFee: sdk.NewDecWithPrec(1, 2),
+	}, []types.PoolAsset{{
+		Weight: sdk.NewInt(100),
+		Token:  sdk.NewCoin("foo", sdk.NewInt(10000)),
+	}, {
+		Weight: sdk.NewInt(100),
+		Token:  sdk.NewCoin("bar", sdk.NewInt(10000)),
+	}}, "")
+	require.NoError(t, err)
+
+	_, err = app.GAMMKeeper.CreatePool(ctx, acc1, types.PoolParams{
+		SwapFee: sdk.NewDecWithPrec(1, 2),
+		ExitFee: sdk.NewDecWithPrec(1, 2),
+	}, []types.PoolAsset{{
+		Weight: sdk.NewInt(70),
+		Token:  sdk.NewCoin("foo", sdk.NewInt(10000)),
+	}, {
+		Weight: sdk.NewInt(100),
+		Token:  sdk.NewCoin("bar", sdk.NewInt(10000)),
+	}}, "")
+	require.NoError(t, err)
+
+	genesis := gamm.ExportGenesis(ctx, app.GAMMKeeper)
+	require.Equal(t, genesis.NextPoolNumber, uint64(3))
+	require.Len(t, genesis.Pools, 2)
+}
+
+func TestMarshalUnmarshalGenesis(t *testing.T) {
+	app := simapp.Setup(false)
+	ctx := app.BaseApp.NewContext(false, tmproto.Header{})
+
+	encodingConfig := simapp.MakeEncodingConfig()
+	appCodec := encodingConfig.Marshaler
+	am := gamm.NewAppModule(appCodec, app.GAMMKeeper, app.AccountKeeper, app.BankKeeper)
+	acc1 := sdk.AccAddress(ed25519.GenPrivKey().PubKey().Address().Bytes())
+	app.BankKeeper.SetBalances(ctx, acc1, sdk.Coins{
+		sdk.NewInt64Coin("foo", 100000),
+		sdk.NewInt64Coin("bar", 100000),
+	})
+
+	_, err := app.GAMMKeeper.CreatePool(ctx, acc1, types.PoolParams{
+		SwapFee: sdk.NewDecWithPrec(1, 2),
+		ExitFee: sdk.NewDecWithPrec(1, 2),
+	}, []types.PoolAsset{{
+		Weight: sdk.NewInt(100),
+		Token:  sdk.NewCoin("foo", sdk.NewInt(10000)),
+	}, {
+		Weight: sdk.NewInt(100),
+		Token:  sdk.NewCoin("bar", sdk.NewInt(10000)),
+	}}, "")
+	require.NoError(t, err)
+
+	genesis := am.ExportGenesis(ctx, appCodec)
+	assert.NotPanics(t, func() {
+		am.InitGenesis(ctx, appCodec, genesis)
+	})
```

Comment:
Can we also have a consistency test here to ensure that the decoded pool is the same as the originally encoded one?

Reply:
We can, just to confirm, you can check v1.0.4 to see more detailed test, https://github.com/osmosis-labs/osmosis/pull/371/files

Reply:
oh perfect, thanks